lib/instrument/index.js

```diff
@@ -0,0 +1,142 @@
+// Adapted from https://github.com/YueChen-C/py-ios-device
+
+import {BaseServiceSocket} from '../base-service';
+import {DTXMessage} from './headers';
+import {DTXEncoder} from './transformer/dtx-encode';
+import {DTXDecoder} from './transformer/dtx-decode';
+import events from 'events';
+import B from 'bluebird';
+const CHECK_FREQ_MS = 100;
+
+const INSTRUMENT_SERVICE_NAME_VERSION_14 = 'com.apple.instruments.remoteserver.DVTSecureSocketProxy';
+const INSTRUMENT_SERVICE_NAME = 'com.apple.instruments.remoteserver';
+
+export const SERVICE = Object.freeze({
+  DEVICE_INFO: 'com.apple.instruments.server.services.deviceinfo',
+  PROCESS_CONTROL: 'com.apple.instruments.server.services.processcontrol',
+  SYSMONTAP: 'com.apple.instruments.server.services.sysmontap',
+  NETWORKING: 'com.apple.instruments.server.services.networking',
+  MOBILE_NOTIFICATIONS: 'com.apple.instruments.server.services.mobilenotifications',
+  GRAPHICS_OPENGL: 'com.apple.instruments.server.services.graphics.opengl',
+  APPLICATION_LISTING: 'com.apple.instruments.server.services.device.applictionListing',
+  CONDITION_INDUCER: 'com.apple.instruments.server.services.ConditionInducer'
+});
+
+function defaultDict (createValue) {
+  return new Proxy(Object.create(null), {
+    get (storage, property) {
+      if (!(property in storage)) {storage[property] = createValue(property);}
+      return storage[property];
+    }
+  });
+}
+
+
+class InstrumentService extends BaseServiceSocket {
+  constructor (socketClient, event) {
+    super(socketClient);
+    this._undefinedCallback = event;
+    this._callbacks = new events.EventEmitter();
+    this._channelCallbacks = new events.EventEmitter();
+    this._replyQueues = defaultDict(function () {
+      return [];
+    });
+    this._channels = {};
+    this._nextIdentifier = 1;
+    this._encoder = new DTXEncoder();
+    this._encoder.pipe(this._socketClient);
+    this._assignClientFailureHandlers(this._encoder);
+    this._decoder = new DTXDecoder();
+    this._socketClient.pipe(this._decoder);
+    this._decoder.on('data', this._receiver.bind(this));
+  }
+  registerSelectorCallback (selector, event) {
+    this._callbacks.addListener(selector, event);
+  }
+  registerUndefinedCallback (event) {
+    this._undefinedCallback = event;
+  }
+  async registerChannelCallback (channel, event) {
+    const channelId = await this.makeChannel(channel);
+    this._channelCallbacks.addListener(channelId, event);
+  }
+  async makeChannel (channel) {
+    if (channel in this._channels) {
+      return this._channels[channel];
+    }
+    const channelID = Object.keys(this._channels).length + 1;
+    await this._call(true, 0, '_requestChannelWithCode:identifier:', channelID, channel);
+    this._channels[channel] = channelID;
+    return channelID;
+  }
+  async waitReply (messageId, timeout = 10000) {
+    return await new B((resolve, reject) => {
```

Comment:
we could use async methods imported from `asyncbox`, for example:

```
import { waitForCondition } from 'asyncbox';

...

return await waitForCondition(async () => {
        // returning true would exit the loop
        // exceptions thrown here are coming through
        return true;
      }, {waitMs: 10000, intervalMs: 500});
```

Reply:
This's great